// @bearer/node-agent/lib/new-rules/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogLevel = exports.RetryType = exports.RemediationType = exports.FilterSetOperator = exports.HttpMethod = exports.FilterType = exports.StageType = void 0;
var StageType;
(function (StageType) {
    StageType["Connect"] = "ConnectStage";
    StageType["Request"] = "RequestStage";
    StageType["Response"] = "ResponseStage";
    StageType["Bodies"] = "BodiesStage";
})(StageType = exports.StageType || (exports.StageType = {}));
var FilterType;
(function (FilterType) {
    // Structural
    FilterType["Not"] = "NotFilter";
    FilterType["FilterSet"] = "FilterSet";
    // Connection
    FilterType["Domain"] = "DomainFilter";
    // Request
    FilterType["HttpMethod"] = "HttpMethodFilter";
    FilterType["Path"] = "PathFilter";
    FilterType["Param"] = "ParamFilter";
    FilterType["RequestHeader"] = "RequestHeaderFilter";
    // Response
    FilterType["ResponseHeader"] = "ResponseHeaderFilter";
    FilterType["StatusCode"] = "StatusCodeFilter";
    // Bodies
    FilterType["RequestBody"] = "RequestBodyFilter";
    FilterType["RequestBodySize"] = "RequestBodySizeFilter";
    FilterType["ResponseBody"] = "ResponseBodyFilter";
    FilterType["ResponseBodySize"] = "ResponseBodySizeFilter";
    // Error
    FilterType["ConnectionError"] = "ConnectionErrorFilter";
    // Bodies/Error
    FilterType["Duration"] = "DurationFilter";
})(FilterType = exports.FilterType || (exports.FilterType = {}));
var HttpMethod;
(function (HttpMethod) {
    HttpMethod["Get"] = "GET";
    HttpMethod["Post"] = "POST";
    HttpMethod["Patch"] = "PATCH";
    HttpMethod["Put"] = "PUT";
    HttpMethod["Delete"] = "DELETE";
    HttpMethod["Head"] = "HEAD";
    HttpMethod["Options"] = "OPTIONS";
})(HttpMethod = exports.HttpMethod || (exports.HttpMethod = {}));
var FilterSetOperator;
(function (FilterSetOperator) {
    FilterSetOperator["All"] = "ALL";
    FilterSetOperator["Any"] = "ANY";
})(FilterSetOperator = exports.FilterSetOperator || (exports.FilterSetOperator = {}));
var RemediationType;
(function (RemediationType) {
    RemediationType["BlockRequest"] = "BlockRequest";
    RemediationType["Timeout"] = "TimeoutRemediation";
    RemediationType["Retry"] = "RetryRemediation";
})(RemediationType = exports.RemediationType || (exports.RemediationType = {}));
var RetryType;
(function (RetryType) {
    RetryType["Exponential"] = "EXPONENTIAL";
    RetryType["Uniform"] = "UNIFORM";
})(RetryType = exports.RetryType || (exports.RetryType = {}));
var LogLevel;
(function (LogLevel) {
    LogLevel["All"] = "ALL";
    LogLevel["Restricted"] = "RESTRICTED";
    LogLevel["Detected"] = "DETECTED";
})(LogLevel = exports.LogLevel || (exports.LogLevel = {}));
